### PR TITLE
[SEDONA-1] Move to jts 1.16.1

### PR DESCRIPTION
**Pull request by nirnir93**
## Is this PR related to a proposed Issue?
#435 #233 

## What changes were proposed in this PR?
Deprecating JTS+ in favor of using JTS as a dependency with wrappers
Updating JTS and GeoTools to latest available versions

## How was this patch tested?
Using the existing tests

## Did this PR include necessary documentation updates?
As far as we could tell, there are none. the changes are internal

## Notes
This is a WIP as there seems to be 2 problems in **python** tests:
1. When projecting a spatial RDD and then calculating it's extent, we get some different results (within 10E-7 of the original test's envelope). We had trouble projecting the envelope manually and so we don't know if this is a more or less accurate result. As far as we could tell, this is caused by the usage of a newer version geotools (23 VS 17)
This effects the following tests:
    - tests/spatial_rdd/test_linestring_rdd.py::TestLineStringRDD::test_constructor
    - tests/spatial_rdd/test_point_rdd.py::TestPointRDD::test_constructor
    - tests/spatial_rdd/test_polygon_rdd.py::TestPolygonRDD::test_constructor
2. In the following tests, the spatial join returns 100 features instead of the expected 200:
    - tests/spatial_operator/test_join_query_correctness.py::TestJoinQueryCorrectness::test_inside_point_join_correctness
    - tests/spatial_operator/test_join_query_correctness.py::TestJoinQueryCorrectness::test_on_boundary_point_join_correctness
    - tests/spatial_operator/test_join_query_correctness.py::TestJoinQueryCorrectness::test_inside_linestring_join_correctness
    - tests/spatial_operator/test_join_query_correctness.py::TestJoinQueryCorrectness::test_overlapped_linestring_join_correctness
    - tests/spatial_operator/test_join_query_correctness.py::TestJoinQueryCorrectness::test_inside_polygon_join_correctness
    - tests/spatial_operator/test_join_query_correctness.py::TestJoinQueryCorrectness::test_overlapped_polygon_join_correctness

Note that for some reason, these tests only fail in python and not in java/scala
We are still working on these issues, and would love if someone has any experience dealing with the mentions tests (@Imbruced )

We decided to open the PR now so that you could approve the design - specifically the wrappers implementation

### CODE REVIEW

**Comment by Imbruced** [comment]
I will look on this, today evening.

File: pom.xml [hunk 1]
```diff
@@ -45,7 +45,8 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <scala.version>2.12.8</scala.version>
         <scala.compat.version>2.12</scala.compat.version>
-        <geotools.version>17.0</geotools.version>
+        <geotools.version>23.2</geotools.version>
+        <jts.version>1.16.1</jts.version>
```
Comment (jnh5y) [inline pom.xml:49]: NB:  JTS 1.17.1 is available.:)  

That said, using the version from GeoTools is probably wise!

**Comment by jiayuasu** [comment]
@Imbruced Hello Pawel, any updates on this?

**Comment by Imbruced** [comment]
@jiayuasu I can see an issue. Python geometry data is translated to jvm using previous serlializer which assume that Geometry has userData attribute. Now it has not and it omits user data. That causing dropping result to 100 pairs (200 expected). I will fix that tomorrow. Example class which causing an issue is **org.datasyslab.geospark.python.translation.HashSetPairRddConverter**

**Comment by Imbruced** [comment]
@jiayuasu I fixed it. But I have some doubts. Did anyone check how this PR affects overall performance ? Do we need this transformations ? 
@nirnir93 can you grant me permissions to your repository ? I have additional branch on which I made required changes to pass all the tests. If you grant me access I will create PR to your branch.

**Comment by nirnir93** [comment]
@Imbruced I've added you

**Comment by nirnir93** [comment]
@Imbruced @jiayuasu 
@netanel246 and I would welcome any input on performance as we tried out best to keep the impact to the necessary minimum (considering some optimizations are no longer possible)

**Comment by Imbruced** [comment]
@nirnir93 I am still not able to push changes. :(

**Comment by Imbruced** [comment]
@nirnir93 I added comments, one file had changes so I added comment within. For 3 python files I can not find appropriate place so I put changes on top of Pr changed files with file names. There were two reasons why all tests did not pass. 
 - First - Wrong classes used, you used jts geoms which did not compare userdata within equals method
- Second - Numerical issue, I did not spot any issue within the code.

**Comment by jiayuasu** [comment]
@Imbruced Hello Pawel, I didn't see your latest update to that "move_to_jts_1.16.1" branch. Where did you push your change to?

To me, the only part could affect the performance is here: https://github.com/deprecatejtsplus/GeoSpark/blob/move_to_jts_1.16.1/core/src/main/java/org/datasyslab/geospark/jts/geom/GeometryFactory.java#L43  Because we added a bunch of if conditions.

What do you mean by "some optimization are no longer possible"?

**Comment by nirnir93** [comment]
@Imbruced 
- I don't know why you can't see your invite, [try this link](https://github.com/deprecatejtsplus/GeoSpark/invitations)
- About the second issue - could it be that geoTools' conversions became more precise? Is there any way to test this conjecture? 

@jiayuasu I mean that (de)serializing manually is not possible without changes to JTS or reflections, so we deleted most of the Serde in favor of Kryo's default behavior. It turned out to still be much better than Java's default serialization mechanism, but somewhat not as good as the manual one

**Comment by jiayuasu** [comment]
@nirnir93 Ok, if so, I think this will significantly decrease our performance. Our existing serializer outperforms Kryo in terms of speed and memory footprint. It can serialize both geometries and spatial indices. This is one of the reason why Sedona beats competitors in many benchmarks.

If there is no other way to bring back the serializer, in the worst case, I will create a fork on LocationTech JTS and periodically and manually pull the latest JTS functions to JTSplus. In this case, no code change is needed for Sedona. But it requires some coding work in the long run and JTS and JTSplus will conflict if they co-exist in one project.

**Comment by nirnir93** [comment]
I'll try to re-add GeometrySerde tomorrow evening

**Comment by nirnir93** [comment]
@jiayuasu Should we revisit the idea of reflections? We can put all the parts pertaining to reflections in a different library (JTSplus) and have it depend on JTS, which would allow for easier JTS updates and should incur fewer conflicts when co-existing with JTS

**Comment by jiayuasu** [comment]
@nirnir93 I think this is a better idea. Do you think this will work? If you need, I will add you and @netanel246  to JTSplus contributors.

**Comment by netanel246** [comment]
@jiayuasu We think it should work. What do you think of creating it as a submodule inside Apache Sedona? this should be more convenient to manage.

**Comment by jnh5y** [comment]
What do you actually need changed in JTS?  I'm a JTS and GeoMesa committer, and I'd be interested in see Sedona work well with both of these projects.

Can you open issues at JTS https://github.com/locationtech/jts/issues, join the JTS Gitter (https://gitter.im/locationtech/jts), or join the JTS-dev list (https://accounts.eclipse.org/mailing-list/jts-dev)?

**Comment by jiayuasu** [comment]
@jnh5y @netanel246 Hey Jim, we created an issue before: https://github.com/locationtech/jts/issues/364

So let me try to create a PR on LocationTech JTS first and see how it goes. If it gets accepted, Sedona will just need to add the support of checking userData attribute.

**Comment by jnh5y** [comment]
@jiayuasu cool.  I'd suggest using an issue or conversation on Gitter first.

If I recall, there are a few things going on.  There was something around Envelope related to R-Trees, right?  That issue may be easy enough to sort out.

For the use of userData, I spoke to Martin about it previously (I poked at integrating GeoMesa and GeoSpark previously).  He has some opinions about how userData should and shouldn't be used. 

That said, I think Sedona could handle the userData bits around JTS Geometries by sending all requests through a custom or specific GeometryFactory...  (I may be misremembering details.)

Generally though, I'm trying to suggest a talk-first pattern to avoid wasted effort;).

**Comment by netanel246** [comment]
@jiayuasu Hi Jia, any updates on this?

**Comment by jiayuasu** [comment]
@netanel246 Sorry, was just busy these days. Over this weekend, I will create a new fork and make a PR on JTS to merge JTSplus functions to JTS. If some JTSplus functions cannot be accepted by JTS, I will simply publish that new fork as a new JTSplus which includes the latest update of JTS.

**Comment by jiayuasu** [comment]
@netanel246 I am currently working on this issue.

I have created a new branch called jtsplus and am implementing my code there. My plan is as follows:

1. Create a git submodule named jts which links to https://github.com/jiayuasu/jts, namely **JTS fork**
2. Sedona Maven project adds the jts/module/core folder as a Maven module. So Sedona can directly use the source files in this folder.
3. Push changes to both Sedona and JTS fork. Several notes:
    1. Move JTSplus codes to Sedona sources as much as possible
    2. Only add the minimum changes to JTS fork
4. Make a PR from JTS fork to LocationTech JTS. Depending on whether the PR can be fully accepted by LocationTech JTS, two options are available:
    1. If the PR is completely accepted, we re-direct the git submodule to LocationTech JTS repo. Once a new JTS release is out in the future, we can remove the git submodule completely and simply add LocationTech JTS as a Maven dependency. Note that, this option most likely won't happen because JTS contributors have clearly said they try to avoid some changes such as leverage "UserData" and change "TreeNode" from private to public.
    2. If the PR is only partially accepted, we can keep the git submodule to JTS fork forever in the future. I will periodically pull the latest change in LocationTech JTS back to JTS fork. In the future, I can publish JTS fork to the Maven Central. But this won't happen until JTS fork reaches a stable status.

This way, 
1. We can easily track the code diff and progresses by comparing JTS fork and LocationTech JTS. 
2. **Another advantage is that this will make the development of Sedona independent on whether the PR gets accepted to LocationTech JTS or not. Currently, the development of Sedona is kind of blocked by this issue and I am keen to fix this ASAP. Once I finish Step 3, the development of Sedona can move on.**
3. If the PR is not accepted, any future users of Sedona can remove the LocationTech JTS packaged by their other dependencies, and simply use JTS fork in the dependency.

I am working on Step 3 right now, it requires some new algorithm work in the JTS STR Tree and Quad Tree because I want to "Only add the minimum changes to JTS fork". This will take a week or two. I will keep you guys posted.

**Pull request closed without merging**